src/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs

```diff
@@ -0,0 +1,298 @@
+﻿// -----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// -----------------------------------------------------------------------
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq;
+
+namespace Microsoft.Internal.Collections
+{
+    internal static partial class CollectionServices
+    {
+        private static readonly Type StringType = typeof(string);
+        private static readonly Type IEnumerableType = typeof(IEnumerable);
+        private static readonly Type IEnumerableOfTType = typeof(IEnumerable<>);
+        private static readonly Type ICollectionOfTType = typeof(ICollection<>);
+
+        public static bool IsEnumerableOfT(Type type)
+        {
+            if (type.IsGenericType)
+            {
+                Type genericType = type.GetGenericTypeDefinition().UnderlyingSystemType;
+
+                if (genericType == IEnumerableOfTType)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public static Type GetEnumerableElementType(Type type)
+        {
+            if (type.UnderlyingSystemType == StringType || !IEnumerableType.IsAssignableFrom(type))
+            {
+                return null;
+            }
+
+            Type closedType;
+            if (ReflectionServices.TryGetGenericInterfaceType(type, IEnumerableOfTType, out closedType))
+            {
+                return closedType.GetGenericArguments()[0];
+            }
+
+            return null;
+        }
+
+        public static Type GetCollectionElementType(Type type)
+        {
+            Type closedType;
+            if (ReflectionServices.TryGetGenericInterfaceType(type, ICollectionOfTType, out closedType))
+            {
+                return closedType.GetGenericArguments()[0];
+            }
+
+            return null;
+        }
+
+        public static ReadOnlyCollection<T> ToReadOnlyCollection<T>(this IEnumerable<T> source)
+        {
+            Assumes.NotNull(source);
+
+            return new ReadOnlyCollection<T>(source.AsArray());
+        }
+
+        public static IEnumerable<T> ConcatAllowingNull<T>(this IEnumerable<T> source, IEnumerable<T> second)
+        {
+            if (second == null || !second.FastAny())
```

Comment:
All this ifs can be simplified to `!second?.FastAny()` maybe in another pass as we want to have the compat package soon.